cleaned release notes final draft (#2689)
kubernetes/sig-release: releases/release-1.32/release-notes/release-notes-draft.md

Reviewer: L19 as well

Reviewer: L128/L129: `Node QueueHint` -> Node's QueueingHint

Reviewer: L183: QueueingHint is no longer alpha, sorry.

Reviewer 2: I think 193 has the PR link. Is this the one you meant? https://github.com/kubernetes/sig-release/blob/2888ce1ccddcb1c493dae6826269363f84c4fbfd/releases/release-1.32/release-notes/release-notes-draft.md?plain=1#L193C3-L193C41

Reviewer 2: > L183: QueueingHint is no longer alpha, sorry.

What's the stage? `inflight_events` looks alpha? https://github.com/kubernetes/kubernetes/pull/127052/files

Reviewer: The original comment means that this metric is for the QHint feature, and QHint is alpha. But, now QHint is beta.
I don't think we need to mention the metric's stage nor QHint's stage, so the current form (no mention to any stages) lgtm.

Reviewer 3: > I think 193 has the PR link

Right, it's then some other formatting issue, probably unnecessary new line, so that it formats as if there was a new paragraph. If it's WAI, then ignore my comment.

Reviewer 4: I would drop lines 1 to 3:
```markdown
## Urgent Upgrade Notes 

### (No, really, you MUST read this before you upgrade)
```

as there aren't any. Maybe:
```markdown
ℹ️ There are no urgent upgrade notes for the v1.32 release
```
? (even better: change the generator post-release to automate this outcome).

Reviewer 5: Yes, I had the same comment. I'm on it.

> ### (No, really, you MUST read this before you upgrade)

Reviewer 2: Should we add https://github.com/kubernetes/kubernetes/issues/129009#issuecomment-2518321729 here? @msau42 is this an upgrade warning?

Reviewer 6: IMHO: not an upgrade warning, CSI driver developers's _own_ binaries may or may not switch to go 1.23 and that's not part of upgrading a cluster. This is comparable to the scheduler plugin changes. (ref https://github.com/kubernetes/kubernetes/pull/129126)

Reviewer 6: IE: these are warnings for third party tools to consider, as a user upgrading a cluster you should wait until your drivers, custom schedulers, etc advertise 1.32 compatibility. As part of those tools implementing 1.32, they should consider these warnings. Versus something like a breaking change you must handle in one of the core components (like removal of dockershim).

Reviewer 5: We've decided on not adding it as a Known Issue https://kubernetes.slack.com/archives/C2C40FMNF/p1733579835815919?thread_ts=1733471881.209099&channel=C2C40FMNF&message_ts=1733579835.815919 .

### API Change

- ([#126620](https://github.com/kubernetes/kubernetes/pull/126620), [@yunwang0911](https://github.com/yunwang0911))

> - ACTION REQUIRED for custom scheduler plugin developers:

Reviewer 6: should we do like https://github.com/kubernetes/kubernetes/pull/129126#discussion_r1877954146 ? (bold ACTION REQUIRED)

### API Change

> - ([#126620](https://github.com/kubernetes/kubernetes/pull/126620), [@yunwang0911](https://github.com/yunwang0911))

Reviewer 3: Description missing

- Graduated Kubelet Memory Manager to GA. ([#128517](https://github.com/kubernetes/kubernetes/pull/128517), [@Tal-or](https://github.com/Tal-or))
- Graduated `SchedulerQueueingHints` to beta; the feature gate is now enabled by default. ([#128472](https://github.com/kubernetes/kubernetes/pull/128472), [@sanposhiho](https://github.com/sanposhiho)) [SIG Scheduling]
- Graduated the `WatchList` feature gate to Beta for kube-apiserver and enabled `WatchListClient` for KCM. ([#128053](https://github.com/kubernetes/kubernetes/pull/128053), [@p0lyn0mial](https://github.com/p0lyn0mial)) [SIG API Machinery and Testing]
- Implemented a queueing hint for PersistentVolumeClaim/Add event in the `CSILimit` plugin. ([#124703](https://github.com/kubernetes/kubernetes/pull/124703), [@utam0k](https://github.com/utam0k)) [SIG Scheduling and Storage]
- Implemented new cluster events `UpdatePodSchedulingGatesEliminated` and `UpdatePodTolerations` for scheduler plugins. ([#127083](https://github.com/kubernetes/kubernetes/pull/127083), [@sanposhiho](https://github.com/sanposhiho))

> - Improved Node's QueueingHint in the `NodeAffinity` plugin by ignoring unrelated changes that keep pods unschedulable. ([#127444](https://github.com/kubernetes/kubernetes/pull/127444), [@dom4ha](https://github.com/dom4ha)) [SIG Scheduling and Testing]
> - Improved `Node QueueHint` in the `NodeResource Fit` plugin by ignoring unrelated changes that keep pods unschedulable. ([#127473](https://github.com/kubernetes/kubernetes/pull/127473), [@dom4ha](https://github.com/dom4ha)) [SIG Scheduling and Testing]

Reviewer 7: I think L127 format should match L126 and plugin's name be fixed:
```suggestion
- Improved Node's QueueingHint in the `NodeAffinity` plugin by ignoring unrelated changes that keep pods unschedulable. ([#127444](https://github.com/kubernetes/kubernetes/pull/127444), [@dom4ha](https://github.com/dom4ha)) [SIG Scheduling and Testing]
- Improved Node's QueueingHint in the `NodeResourceFit` plugin by ignoring unrelated changes that keep pods unschedulable. ([#127473](https://github.com/kubernetes/kubernetes/pull/127473), [@dom4ha](https://github.com/dom4ha)) [SIG Scheduling and Testing]
```

- Fixed documentation for the `apiserver_admission_webhook_fail_open_count` and `apiserver_admission_webhook_request_total` metrics. The `type` label can have a value of "admit", not "mutating". ([#127898](https://github.com/kubernetes/kubernetes/pull/127898), [@modulitos](https://github.com/modulitos))
- Kubeadm: fixed a misleading output (typo) about control-plane joining instructions when executing the "kubeadm init" command. ([#128118](https://github.com/kubernetes/kubernetes/pull/128118), [@amaddio](https://github.com/amaddio))
- The kubelet, when using `--cloud-provider=external` can use the `--node-ip` flag with one of the unspecified addresses 0.0.0.0 or ::, to create the Node with the IP of the default gateway of the corresponding IP family and then delegating the responsibility to the external cloud provider. This solve the bootstrap problems of out of tree cloud providers that are deployed as Pods within the cluster. ([#125337](https://github.com/kubernetes/kubernetes/pull/125337), [@aojea](https://github.com/aojea)) [SIG Cloud Provider, Network, Node and Testing]

> - Request header UID propagation is gated behind an alpha RemoteRequestHeaderUID feature gate ([#129081](https://github.com/kubernetes/kubernetes/pull/129081), [@stalz](https://github.com/stlaz)) [SIG API SIG API Machinery, cluster lifecycle, testing]

Reviewer 2: ```suggestion
- Request header UID propagation is gated behind an Alpha RemoteRequestHeaderUID feature gate. ([#129081](https://github.com/kubernetes/kubernetes/pull/129081), [@stalz](https://github.com/stlaz)) [SIG API SIG API Machinery, cluster lifecycle, testing]
```

- Kubeadm: fixed a misleading output (typo) about control-plane joining instructions when executing the "kubeadm init" command. ([#128118](https://github.com/kubernetes/kubernetes/pull/128118), [@amaddio](https://github.com/amaddio))
- The kubelet, when using `--cloud-provider=external` can use the `--node-ip` flag with one of the unspecified addresses 0.0.0.0 or ::, to create the Node with the IP of the default gateway of the corresponding IP family and then delegating the responsibility to the external cloud provider. This solve the bootstrap problems of out of tree cloud providers that are deployed as Pods within the cluster. ([#125337](https://github.com/kubernetes/kubernetes/pull/125337), [@aojea](https://github.com/aojea)) [SIG Cloud Provider, Network, Node and Testing]
- Added request header UID propagation, behind an alpha `RemoteRequestHeaderUID` feature gate. ([#129081](https://github.com/kubernetes/kubernetes/pull/129081), [@stalz](https://github.com/stlaz)) [SIG API SIG API Machinery, cluster lifecycle, testing]

> ### Failing Test

Reviewer 4: (nit) We could fix the capitalization of kubelet in this section.

Reviewer 5: Did it for kubelet and kubeadm throughout.

  https://github.com/kubernetes-sigs/kustomize/releases/tag/kustomize%2Fv5.4.2
  and https://github.com/kubernetes-sigs/kustomize/releases/tag/kustomize%2Fv5.5.0. ([#127965](https://github.com/kubernetes/kubernetes/pull/127965), [@koba1t](https://github.com/koba1t))
- `ComponentSLIs` feature is marked as GA and locked. ([#128317](https://github.com/kubernetes/kubernetes/pull/128317), [@Jefftree](https://github.com/Jefftree)) [SIG Architecture and Instrumentation]
- `kubectl apply --server-side` now supports `--subresource` congruent to `kubectl patch`. ([#127634](https://github.com/kubernetes/kubernetes/pull/127634), [@deads2k](https://github.com/deads2k)) [SIG CLI and Testing]

> - kubelet: fixes an issue mounting CSI volumes on Windows nodes in 1.32.0 release candidates. ([#129083](https://github.com/kubernetes/kubernetes/pull/129083) [liggitt](https://github.com/liggitt)) [SIG API Machinery, architecture, auth, cli, cloud-provider, cluster-lifecycle, instrumentation,network,node, release, storage, windows ]

Reviewer 2: ```suggestion
- kubelet: Fixed an issue mounting CSI volumes on Windows nodes in 1.32.0 release candidates. ([#129083](https://github.com/kubernetes/kubernetes/pull/129083) [liggitt](https://github.com/liggitt)) [SIG API Machinery, architecture, auth, cli, cloud-provider, cluster-lifecycle, instrumentation,network,node, release, storage, windows ]
```
